### PR TITLE
Add support for user/password auth

### PR DESCRIPTION
**Pull request by Spider84**

### CODE REVIEW

File: src/hev-socks5-client.c [hunk 1]
```diff
@@ -110,7 +110,20 @@ hev_socks5_client_write_request (HevSocks5Client *self)
 
     auth.ver = HEV_SOCKS5_VERSION_5;
     auth.method_len = 1;
-    auth.methods[0] = HEV_SOCKS5_AUTH_METHOD_NONE;
+
+    if (self->base.auth.user && self->base.auth.pass) {
+        //Based on https://datatracker.ietf.org/doc/html/rfc1929
+
+        auth.methods[0] = HEV_SOCKS5_AUTH_METHOD_USER;
```
Comment (heiher) [inline src/hev-socks5-client.c:117]: Although the right way is to send username/password when the server choose authentication, but i think this way is better for help hide protocol.

File: src/hev-socks5-client.c [hunk 2]
```diff
@@ -110,7 +110,20 @@ hev_socks5_client_write_request (HevSocks5Client *self)
...
+
+        auth.methods[1] = HEV_SOCKS5_AUTH_VERSION_1;
+        auth.methods[2] = strlen(self->base.auth.user);
+        strncpy((char *)&auth.methods[3], self->base.auth.user, auth.methods[2]);
+        auth.methods[2+auth.methods[2]+1] = strlen(self->base.auth.pass);
+        strncpy((char *)&auth.methods[2+auth.methods[2]+2], self->base.auth.pass, auth.methods[2+auth.methods[2]+1]);
```
Comment (heiher) [inline src/hev-socks5-client.c:123]: Use variable-length iovec instead of strncpy?

File: src/hev-socks5-client.c [hunk 3]
```diff
@@ -148,6 +161,10 @@ hev_socks5_client_write_request (HevSocks5Client *self)
 
     iov[0].iov_base = &auth;
     iov[0].iov_len = 3;
+
+    if (auth.methods[0] == HEV_SOCKS5_AUTH_METHOD_USER)
+        iov[0].iov_len += ((2+auth.methods[2]+2)+auth.methods[2+auth.methods[2]+1]);
```
Comment (heiher) [inline src/hev-socks5-client.c:166]: Initialize at the same time as the data?

```
auth.ver = HEV_SOCKS5_VERSION_5;
auth.method_len = 1;
...
iov[0].iov_base = &auth;
iov[0].iov_len = ...;
```

**Review by heiher — CHANGES_REQUESTED** [review]
Thanks for your patch, It might be better if you can make some changes.

BTW, You should to use `clang-format -i src/*` to keep the code style.

File: src/hev-socks5-client.c (hunk 1 above)
Comment (Spider84) [reply src/hev-socks5-client.c:117]: This is required by RFC. You know any other methods?

File: src/hev-socks5-client.c (hunk 2 above)
Comment (Spider84) [reply src/hev-socks5-client.c:123]: ?

File: src/hev-socks5-client.c (hunk 3 above)
Comment (Spider84) [reply src/hev-socks5-client.c:166]: I just trying to save you code-line sequence

File: src/hev-socks5-client.c (hunk 1 above)
Comment (heiher) [reply src/hev-socks5-client.c:117]: It seem okay, when the client only provides one method for authentication.

File: src/hev-socks5-client.c (hunk 2 above)
Comment (heiher) [reply src/hev-socks5-client.c:123]:
```c
struct iovec iov[6];
int i = 0;

iov[i].iov_base = &auth;
iov[i++].iov_len = 3;

if (self->base.auth.user && self->base.auth.pass) {
    auth.methods[0] = HEV_SOCKS5_AUTH_METHOD_USER;
    auth.methods[1] = HEV_SOCKS5_AUTH_VERSION_1;
    auth.methods[2] = strlen (self->base.auth.user);
    auth.methods[3] = strlen (self->base.auth.pass);
    iov[i].iov_base = self->base.auth.user;
    iov[i++].iov_len = auth.methods[2];
    iov[i].iov_base = &auth.methods[3];
    iov[i++].iov_len = 1;
    iov[i].iov_base = self->base.auth.user;
    iov[i++].iov_len = auth.methods[3];
} else {
    auth.methods[0] = HEV_SOCKS5_AUTH_METHOD_NONE;
}

iov[i].iov_base = &req;
iov[i++].iov_len = 3;
...

memset (&mh, 0, sizeof (mh));
mh.msg_iov = iov;
mh.msg_iovlen = i;
```

File: src/hev-socks5-client.c (hunk 1 above)
Comment (Spider84) [reply src/hev-socks5-client.c:117]: Already - two :) NONE and USER!

File: src/hev-socks5-client.c (hunk 2 above)
Comment (Spider84) [reply src/hev-socks5-client.c:123]: О_о

Comment (heiher) [reply src/hev-socks5-client.c:123]: The goal is reduce memory copy. ;-D

**Comment by heiher** [comment]
Merged with amend.

Thank you.

**Pull request closed without merging**